include/pinocchio/parsers/graph/model-graph.hpp

```diff
@@ -0,0 +1,391 @@
+
+      /// @brief boolean to know if we are in a forward or backward edge
+      bool forward = true;
+    };
+
+    /// @brief Contains information about how \ref buildModel walked the \ref ModelGraph to
+    /// construct a \ref Model.
+    /// All members are considered internal.
+    struct ModelGraphBuildInfo
+    {
+      /// Map joint name to joint direction.
+      std::unordered_map<std::string, bool> _joint_forward;
+      /// True if the root joint is fixed.
+      bool _is_fixed;
+    };
+
+    /// @brief Represents multibody model as a bidirectional graph.
+    ///
+    /// This is an intermediate step before creating a model, that
+    /// allows more flexibility as to which body will be the root...
+    struct PINOCCHIO_PARSERS_DLLAPI ModelGraph
+    {
+      typedef boost::
+        adjacency_list<boost::vecS, boost::vecS, boost::directedS, ModelGraphVertex, ModelGraphEdge>
+          Graph;
+      typedef typename boost::graph_traits<Graph>::vertex_descriptor VertexDesc;
+      typedef typename boost::graph_traits<Graph>::edge_descriptor EdgeDesc;
+
+      ModelGraph() = default;
+      /// \brief Add a new vertex to the graph
+      ///
+      /// \param[in] vertex_name Name of the vertex
+      /// \param[in] frame which type of frame will be added to the model (op_frame, sensor, body)
+      void addFrame(const std::string & vertex_name, const FrameVariant & frame);
+
+      /// \brief Add a new body to the graph
+      ///
+      /// \param[in] vertex_name Name of the vertex
+      /// \param[in] inert inertia of the body
+      void addBody(const std::string & vertex_name, const Inertia & inert);
+
+      GeometryBuilder useGeometryBuilder();
+      void addGeometry(const std::string & vertex_name, const Geometry & geom);
+      void addGeometries(const std::string & vertex_name, const std::vector<Geometry> & geoms);
+
+      /// \brief Add edges (joint) to the graph. Since it's a bidirectional graph,
+      /// edge and its reverse are added to the graph.
+      ///
+      /// \param[in] joint_name Name of the edge
+      /// \param[in] joint Type of the joint
+      /// \param[in] source_body Vertex that is supporting the edge
+      /// \param[in] source_to_joint Transformation from supporting vertex to edge
+      /// \param[in] target_body Vertex that is supported by edge
+      /// \param[in] joint_to_target Transformation from edge to supported vertex
+      /// \param[in] q_ref q offset of the joint
+      ///
+      /// \note Since it's a bidirectional graph, two edges are added to the graph.
+      /// Joints and transformation are inverted, to create reverse edge.
+      void addJoint(
+        const std::string & joint_name,
+        const JointVariant & joint,
+        const std::string & source_body,
+        const SE3 & source_to_joint,
+        const std::string & target_body,
+        const SE3 & joint_to_target,
+        const boost::optional<Eigen::VectorXd> & q_ref = boost::none);
```

Comment:
```suggestion
      void addJoint(
        const std::string & joint_name,
        const JointVariant & joint,
        const std::string & source_body,
        const SE3 & source_to_joint,
        const std::string & target_body,
        const SE3 & joint_to_target);
```

On this version, only use the minimal parameter set.